examples/datalog/stratifier.flix

```diff
@@ -0,0 +1,99 @@
+        let removeBool = triple -> match triple {case (a, _, b) => (a, b)};
+        let facts = project map(removeBool, pos), map(removeBool, neg) into PositiveDependencyEdge, NegativeDependencyEdge;
+        let rules = #{
+            Stratum(pd; 1) :- PositiveDependencyEdge(pd, _). Stratum(pd; 1) :- PositiveDependencyEdge(_, pd).
+            Stratum(pd; 1) :- NegativeDependencyEdge(pd, _). Stratum(pd; 1) :- NegativeDependencyEdge(_, pd).
+            Stratum(ph; max(pbs, phs)) :- PositiveDependencyEdge(ph, pb), Stratum(pb; pbs), Stratum(ph; phs).
+            Stratum(ph; max(pbs + 1, phs)) :-
+                NegativeDependencyEdge(ph, pb),
+                Stratum(pb; pbs),
+                Stratum(ph; phs),
+                if pbs <= numberOfPredicates.
+                // allow one level of strata above the bound for a stratification check later
+        };
+        let solution = solve facts, rules;
+        let stratified = (query solution select () from Stratum(_; s) where s >= numberOfPredicates + 1) |> Array.isEmpty;
+        if (not stratified)
+            None
+        else
+            let mappings = query solution select (pd, s) from Stratum(pd; s);
+            Some(Array.toMap(mappings))
+
+    pub def testScenario01(): Unit & Impure =
+        println("testScenario01");
+        // A :- B, C
+        // B :- D
+        // C :- not D
+        let p = Program(
+            Constraint(HeadAtom(Atom(PredicateSymbol("A"))), Positive(Atom(PredicateSymbol("B"))) :: Positive(Atom(PredicateSymbol("C"))) :: Nil) ::
+            Constraint(HeadAtom(Atom(PredicateSymbol("B"))), Positive(Atom(PredicateSymbol("D"))) :: Nil) ::
+            Constraint(HeadAtom(Atom(PredicateSymbol("C"))), Negative(Atom(PredicateSymbol("D"))) :: Nil) ::
+            Nil
+        );
+        let result = ullman(p, 4);
+        let expected = "Map#{PredicateSymbol(A) -> 2, PredicateSymbol(B) -> 1, PredicateSymbol(C) -> 2, PredicateSymbol(D) -> 1}";
+        outputString(result, expected) |> println
+
+    pub def testScenario02(): Unit & Impure =
+        println("testScenario02");
+        // A :- B, C
+        // B :- D
+        // C :- not D
+        // D :- A
+        let p = Program(
+            Constraint(HeadAtom(Atom(PredicateSymbol("A"))), Positive(Atom(PredicateSymbol("B"))) :: Positive(Atom(PredicateSymbol("C"))) :: Nil) ::
+            Constraint(HeadAtom(Atom(PredicateSymbol("B"))), Positive(Atom(PredicateSymbol("D"))) :: Nil) ::
+            Constraint(HeadAtom(Atom(PredicateSymbol("C"))), Negative(Atom(PredicateSymbol("D"))) :: Nil) ::
+            Constraint(HeadAtom(Atom(PredicateSymbol("D"))), Negative(Atom(PredicateSymbol("A"))) :: Nil) ::
+            Nil
+        );
+        let result = ullman(p, 4);
+        let expected = "Not Stratified";
+        outputString(result, expected) |> println
+
+    def outputString(value: Option[Map[PredicateSymbol, Int]], expected: String): String =
```

Comment:
Drop this function. Remember this is meant to be _an example_ not to be some testing infrastructure thing.